rasa/core/policies/rule_policy.py

```diff
@@ -74,6 +79,183 @@ def _create_feature_key(self, states: List[Dict]) -> Text:
 
         return feature_str
 
+    @staticmethod
+    def _get_active_form_name(state: Dict[Text, float]) -> Optional[Text]:
+        # by construction there is only one active form
+        return next(
+            (
+                state_name[len(ACTIVE_FORM_PREFIX) :]
+                for state_name, prob in state.items()
+                if ACTIVE_FORM_PREFIX in state_name
+                and state_name != ACTIVE_FORM_PREFIX + "None"
+                and prob > 0
+            ),
+            None,
+        )
+
+    @staticmethod
+    def _prev_action_listen_in_state(state: Dict[Text, float]) -> bool:
+        return any(
+            PREV_PREFIX + ACTION_LISTEN_NAME in state_name and prob > 0
+            for state_name, prob in state.items()
+        )
+
+    @staticmethod
+    def _modified_states(
+        states: List[Dict[Text, float]]
+    ) -> List[Optional[Dict[Text, float]]]:
+        """Modifies the states to create feature keys for form unhappy path conditions.
+
+        Args:
+            states: a representation of a tracker
+                as a list of dictionaries containing features
+
+        Returns:
+            modified states
+        """
+
+        indicator = PREV_PREFIX + RULE_SNIPPET_ACTION_NAME
+        state_only_with_action = {indicator: 1}
+        # leave only last 2 dialogue turns to
+        # - capture previous meaningful action before action_listen
+        # - ignore previous intent
+        if len(states) > 2 and states[-2] is not None:
+            state_only_with_action = {
+                state_name: prob
+                for state_name, prob in states[-2].items()
+                if PREV_PREFIX in state_name and prob > 0
+            }
+
+        # add `prev_...` to show that it should not be a first turn
+        if indicator not in state_only_with_action and indicator not in states[-1]:
+            return [{indicator: 1}, state_only_with_action, states[-1]]
+
+        return [state_only_with_action, states[-1]]
+
+    @staticmethod
+    def _clean_feature_keys(lookup: Dict[Text, Text]) -> Dict[Text, Text]:
+        # remove action_listens that were added after conditions
+        updated_lookup = lookup.copy()
+        for feature_key, action in lookup.items():
+            # Delete rules if there is no prior action or if it would predict
+            # the `...` action
+            if PREV_PREFIX not in feature_key or action == RULE_SNIPPET_ACTION_NAME:
+                del updated_lookup[feature_key]
+            elif RULE_SNIPPET_ACTION_NAME in feature_key:
+                # If the previous action is `...` -> remove any specific state
+                # requirements for that state (anything can match this state)
+                new_feature_key = re.sub(
+                    rf".*{PREV_PREFIX}\.\.\.[^|]*", "", feature_key
+                )
+
+                if new_feature_key:
+                    if new_feature_key.startswith("|"):
```

Comment:
What does `|` mean again? Does it make sense to replace it with a constant?

Reply:
not yet, this will be the part of next refactor after e2e